trapper/data/metadata_handlers/metadata_handler.py

```diff
@@ -0,0 +1,120 @@
+import logging
+from typing import Optional, Tuple, Union
+
+import numpy as np
+
+from trapper.common import Registrable
+from trapper.data import IndexedInstance
+from trapper.data.label_mapper import LabelMapper
+from trapper.data.tokenizers import TokenizerWrapper
+
+logger = logging.getLogger(__file__)
+
+
+class MetadataHandler(Registrable):
+    """
+    This callable class is responsible for postprocessing evaluation output
+    :py:class:`transformers.EvalPrediction` used in
+    :py:class:`trapper.training.TransformerTrainer`. It is used to convert to
+    suitable evaluation format for the specified metrics before metric computation.
+    Do not override the `__call__()` method, instead override `postprocess()`
+    methods to your task's needs and also override `extract_metadata()`
+    if the task needs metadata to postprocess the evaluation output. See
+    `MetadataHandlerForQuestionAnswering` for an example.
+
+    Args:
+        tokenizer_wrapper (): Required to postprocess eval outputs of a model.
+        label_mapper (): Only used in some tasks that require mapping between
+            categorical labels and integer ids such as token classification.
+    """
+
+    default_implementation = "default"
+
+    def __init__(
+        self,
+        tokenizer_wrapper: TokenizerWrapper,
+        label_mapper: Optional[LabelMapper] = None,
+    ):
+        self._tokenizer = tokenizer_wrapper
+        self._label_mapper = label_mapper
+
+    @property
+    def tokenizer(self):
+        return self._tokenizer.tokenizer
+
+    @tokenizer.setter
+    def tokenizer(self, value):
+        if isinstance(value, TokenizerWrapper):
+            self._tokenizer = value
+        else:
+            raise ValueError(
+                f"The value must be an instance of a "
+                f"class derived from {TokenizerWrapper}"
+            )
+
+    @property
+    def label_mapper(self):
+        return self._label_mapper
+
+    @label_mapper.setter
+    def label_mapper(self, value):
+        if isinstance(value, LabelMapper):
+            self._label_mapper = value
+        else:
+            raise ValueError(
+                f"The value must be an instance of a "
+                f"class derived from {LabelMapper}"
+            )
+
+    def __call__(self, instance: IndexedInstance, split: str) -> IndexedInstance:
+        """
+        Where extraction from instances happen through an abstractmethod
+        extract_metadata, returns the input instance as is. This method cannot be
+        overridden in child classes.
+
+        Args:
+            instance: Indexed instance.
+            split: split of the data.
+
+        Returns: IndexedInstance as is.
+        """
+        # Currently through HF trainer validation split is used for eval
+        if split == "validation":
+            self.extract_metadata(instance)
+        return instance
+
+    def extract_metadata(self, instance: IndexedInstance) -> None:
```

Comment:
in order to use default metadata handler as is, this is the only reason for not making this method abstract.

Reply:
Well, there is another problem with this. Now, by default, we pass a callable that returns none to the dataset.map by this statement `        dataset.map(self.extract_metadata)`. I think, it should return the sample as it is. Moreover, that statement in statement in the _call_ method might be dangerous. You are relying on the self.extract_metadata function to have side effects that does what you think it should do. However, it would be cleaner if we let extract_metadata to always return a type what we expect (if there is any) and do `return dataset.map(self.extract_metadata)` in the _call_ method. However, I am not that familiar to these classes, so I am open for criticism.

Reply:
Here, the intended usage is to have a single `__call__()` method which should not be overridden in any child class (as we discussed before typing.final does support this for python 3.8, but currently no option for that). The intention of the existence of `extract_metadata` is due to to main purpose:
 
1. Provide an interface to `__call__()` method to access dataset instances, and store the required information. Since `__call__()` should not be overridden, we provide an overridable method for child classes.

> However, it would be cleaner if we let extract_metadata to always return a type what we expect (if there is any) and do return dataset.map(self.extract_metadata) in the call method.

2. Second purpose is actually providing us what you mentioned here. If we specify a return type and return an instance we can never guarantee that the instance is manipulated or changed (in any ways). We do not want to have any manipulation on `dataset` passed to `self.extract_metadata()`. Thus, specifying return type as `None` guarantees that `dataset` is not manipulated in any way by the map function, making `self.extract_metadata()` read-only, without a write access. In this method user only extract the metadata required for postprocess of prediction outputs and store them in object variables (attributes), for example like in question answering case.